lib/voicemail-admin.js

```diff
@@ -0,0 +1,771 @@
+/**
+ * Voicemail Administrator application.
+ *
+ * @module voicemail-admin
+ * @copyright 2014, Digium, Inc.
+ * @license Apache License, Version 2.0
+ * @author Jonathan R. Rose <jrose@digium.com>
+ */
+
+'use strict';
+
+var menu = require('../menu.json');
+var prompt = require('prompt');
+
+var dal;
+var logger;
+
+var sprintf = require('sprintf').sprintf;
+
+var commands = createApi();
+
+var Q = require('q');
+
+prompt.message = '';
+prompt.delimiter = '';
+var inputIndicator = {
+  name: 'command',
+  message: '=>'.green
+};
+
+var internalCommands = ['exit', 'help'];
+
+function processOption(result) {
+  var entryComponents;
+  var baseCommand;
+  var chosenCommand = false;
+  var functionPtr;
+  var index;
+
+  if (!result) {
+    return Q.fcall(exit);
+  }
+
+  /* Split the string by spaces, but keep stuff between quotes together */
+  entryComponents = result.command.match(/(?:[^\s"]+|"[^"]*")+/g);
+  entryComponents = entryComponents.map(function(component) {
+    return component.replace(/"/g, '');
+  });
+
+  if (!entryComponents) {
+    return Q.fcall(function () {
+      /* do nothing */
+    });
+  }
+
+  /* help is the only single word command -- everything else is two words */
+  if (entryComponents.length === 1 ||
+      internalCommands.indexOf(entryComponents[0]) > -1) {
+    baseCommand = entryComponents[0];
+  } else {
+    baseCommand = entryComponents[0] + ' ' + entryComponents[1];
+  }
+
+  for (index in menu.commands) {
+    if (baseCommand === menu.commands[index].command) {
+      chosenCommand = menu.commands[index];
+      break;
+    }
+  }
+
+  if (!chosenCommand) {
+    return invalidCommand(result);
+  }
+
+  functionPtr = commands[chosenCommand.action];
+
+  return functionPtr(chosenCommand, entryComponents);
+}
+
+var getPrompt = Q.denodeify(prompt.get);
+
+function invalidCommand(result) {
+  return Q.fcall(function () {
```

Comment:
It looks like fcall is used a lot to turn non IO bound functions to promises. Is there a reason for this? If a function doesn't have any async IO in it, I'm not sure why you would need to wrap it in a promise.